hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala

```diff
@@ -5,35 +5,18 @@ import is.hail.utils.HailException
 object FoldConstants {
   def apply(ir: BaseIR, canGenerateLiterals: Boolean = true): BaseIR =
     RewriteBottomUp(ir, {
-      case ir: IR =>
-        ir match {
-          case _: Ref |
-               _: In |
-               _: ApplySeeded |
-               _: ApplyAggOp |
-               _: ApplyScanOp |
-               _: SeqOp |
-               _: Begin |
-               _: InitOp |
-               _: ArrayRange |
-               _: Die => None
-          case MakeStruct(fields) if fields.isEmpty => None
-          case Let(name, value, body) if IsConstant(value) =>
-            Some(FoldConstants(Subst(body, BindingEnv(Env.empty[IR].bind(name, value))), canGenerateLiterals))
-          case _ if IsConstant(ir) => None
-          case _ if !Interpretable(ir) => None
-          case _ =>
-            if (ir.children.forall {
-              case c: IR => IsConstant(c)
-              case _ => false
-            } && (canGenerateLiterals || CanEmit(ir.typ))) {
-              Some(try {
-                Literal.coerce(ir.typ, Interpret(ir, optimize = false))
-              } catch {
-                case e: HailException => Die(e.getMessage + e.getStackTrace.map(s => s"\n    $s").mkString(""), ir.typ)
-              })
-            }
-            else None
+      case ir: IR if !IsConstant(ir) &&
```

Comment:
@tpoterba are you OK with this?  FoldConstants was an insane mess and I cleaned it up.

Reply:
won't this run `Die(Str(...))` on every pass of the optimizer? That seems kind of bad. Should we not wrap this in a die node?

Reply:
Sorry, I put back the set of nodes we shouldn't try to fold (including Die).

Reply:
I think your solution was simpler, we just need to think through the edge cases and make sure they're handled in the base case.